### PR TITLE
8265325: Optimize StubRoutines::dpow() for Math.pow(x, 0.5)

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

I'd like to optimize the StubRoutines::dpow() for Math.pow(x, 0.5).

In the pow and sqrt discussion [1], Joe taught me that the Java library implementation of pow has been optimized for pow(x, 2.0) [2] and pow(x, 0.5) [3].
However, the hotspot StubRoutines::dpow() only implements the same opt for pow(x, 2.0), but still not for pow(x, 0.5).
This patch optimizes StubRoutines::dpow() for pow(x, 0.5).

Although not all Math.pow(x, 0.5) can be replaced with sqrt(x), we can still do it safely for the following cases:
  1) x >= 0.0    (fully implemented)
  2) x is +Inf   (fully implemented)
  3) x is NaN    (can be further divided into +NaN and -NaN and only +NaN is implemented)

The effect of this opt has been tested on serveral platforms showing 3.0x ~ 6.3x performance improvement.
And no performance drop was observed.

Testing:
  - tier1 ~ tier3 on Linux/x64

Thanks.
Best regards,
Jie

[1] https://mail.openjdk.java.net/pipermail/core-libs-dev/2021-April/076220.html
[2] https://github.com/openjdk/jdk/blob/d84a7e55be40eae57b6c322694d55661a5053a55/src/java.base/share/classes/java/lang/FdLibm.java#L362
[3] https://github.com/openjdk/jdk/blob/d84a7e55be40eae57b6c322694d55661a5053a55/src/java.base/share/classes/java/lang/FdLibm.java#L364

Detailed performance numbers:
* Linux/Intel
```
--------- Before -----------
Benchmark                      (seed)   Mode  Cnt       Score       Error   Units
MathBench.powDouble                 0  thrpt    8  218783.605 ?   838.379  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8   45498.351 ?     7.558  ops/ms
MathBench.powDouble0Dot5Const       0  thrpt    8   45243.530 ?  1097.100  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.031 ?     0.001  ops/ms
MathBench.powDoubleLoop             0  thrpt    8       0.031 ?     0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8  176106.602 ? 13127.650  ops/ms
----------------------------

--------- After -----------
Benchmark                      (seed)   Mode  Cnt       Score       Error   Units
MathBench.powDouble                 0  thrpt    8  219930.462 ?   181.922  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8  204966.834 ?   329.032  ops/ms   <-- 4.5x up
MathBench.powDouble0Dot5Const       0  thrpt    8  203004.302 ?   684.072  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.121 ?     0.001  ops/ms   <-- 3.9x up
MathBench.powDoubleLoop             0  thrpt    8       0.031 ?     0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8  178818.861 ? 16235.465  ops/ms
----------------------------
```

* Linux/AMD
```
--------- Before -----------
Benchmark                      (seed)   Mode  Cnt       Score     Error   Units
MathBench.powDouble                 0  thrpt    8  100741.348 ? 207.766  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8   33896.623 ? 103.352  ops/ms
MathBench.powDouble0Dot5Const       0  thrpt    8   34195.944 ? 230.703  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.039 ?   0.001  ops/ms
MathBench.powDoubleLoop             0  thrpt    8       0.038 ?   0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8   72000.166 ? 135.002  ops/ms
----------------------------

--------- After -----------
Benchmark                      (seed)   Mode  Cnt       Score     Error   Units
MathBench.powDouble                 0  thrpt    8  100738.866 ? 222.820  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8  100799.098 ?  95.537  ops/ms   <-- 3.0x up
MathBench.powDouble0Dot5Const       0  thrpt    8  100765.571 ? 178.436  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.244 ?   0.002  ops/ms   <-- 6.3x up
MathBench.powDoubleLoop             0  thrpt    8       0.038 ?   0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8   71758.725 ? 339.660  ops/ms
----------------------------
```

* MacOS/Intel
```
--------- Before -----------
Benchmark                      (seed)   Mode  Cnt       Score      Error   Units
MathBench.powDouble                 0  thrpt    8  238064.722 ? 5181.318  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8   59235.979 ? 2046.519  ops/ms
MathBench.powDouble0Dot5Const       0  thrpt    8   59695.014 ? 1079.692  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.040 ?    0.001  ops/ms
MathBench.powDoubleLoop             0  thrpt    8       0.041 ?    0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8  238391.026 ? 2743.385  ops/ms
----------------------------

--------- After -----------
Benchmark                      (seed)   Mode  Cnt       Score       Error   Units
MathBench.powDouble                 0  thrpt    8  238582.414 ?  3661.261  ops/ms
MathBench.powDouble0Dot5            0  thrpt    8  224102.701 ?  2846.892  ops/ms   <-- 3.8x up
MathBench.powDouble0Dot5Const       0  thrpt    8  224542.331 ? 19027.596  ops/ms
MathBench.powDouble0Dot5Loop        0  thrpt    8       0.158 ?     0.002  ops/ms   <-- 4.0x up
MathBench.powDoubleLoop             0  thrpt    8       0.041 ?     0.001  ops/ms
StrictMathBench.powDouble         N/A  thrpt    8  233689.504 ? 10141.034  ops/ms
----------------------------
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265325](https://bugs.openjdk.java.net/browse/JDK-8265325): Optimize StubRoutines::dpow() for Math.pow(x, 0.5)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3536/head:pull/3536` \
`$ git checkout pull/3536`

Update a local copy of the PR: \
`$ git checkout pull/3536` \
`$ git pull https://git.openjdk.java.net/jdk pull/3536/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3536`

View PR using the GUI difftool: \
`$ git pr show -t 3536`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3536.diff">https://git.openjdk.java.net/jdk/pull/3536.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp
Comment (vnkozlov) [inline src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2]: Please, don't modify copyright line of other company. Add new line for your company.

File: test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java:41]: `r1` should be static value to avoid compiling `sqrt()` method as intrinsic.

**Review by vnkozlov — COMMENTED** [review]
Thank you for testing all modes.

Note, when StubRoutines::dpow() is not used (-XX:-UseLibmIntrinsic or -XX:DisableIntrinsic=_pow) we use C code which also have this (and others) optimization already:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntimeTrans.cpp#L483

File: test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java
Comment (DamonFool) [reply test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java]: > `r1` should be static value to avoid compiling `sqrt()` method as intrinsic.

Hi @vnkozlov ,

Thanks for your review.

Patch has been updated based on your comments.

To be honest, I didn't get why r1 should be static value.
I think both static and non-static should be OK for the test.

So what would happen is sqrt() is intrinsified?
Could you please make it more clearer?

Thanks.
Best regards,
Jie

Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java]: Sorry, my bad. I thought to calculate `sqrt(a)` outside `test` method so that only `pow()` code is tested.
Like calculating `gold` value not in compiled code. 
But the test have to use sqrt() for each value `a` and `sqrt` is 1 HW instruction so my suggestion was stupid.
Please, revert this change. Original test code was fine.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/intrinsics/math/TestPow0Dot5Opt.java:41]: > Sorry, my bad. I thought to calculate `sqrt(a)` outside `test` method so that only `pow()` code is tested.
> Like calculating `gold` value not in compiled code.
> But the test have to use sqrt() for each value `a` and `sqrt` is 1 HW instruction so my suggestion was stupid.
> Please, revert this change. Original test code was fine.

Thanks for your clarification.
I got your point.

Updated.
Thanks.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by DamonFool** [comment]
> Good.

Thanks @vnkozlov .

May I get a second review for this change?
Or it's already OK to be pushed?
Thanks.

**Comment by neliasso** [comment]
> May I get a second review for this change?
> Or it's already OK to be pushed?

You need a second review for changes to Hotspot. I will provide it.

**Review by neliasso — APPROVED** [review]
Approved.

**Comment by DamonFool** [comment]
> Approved.

Thanks @neliasso .

**Pull request closed without merging**
